### PR TITLE
Clear Python-side native caches before a registry reset

### LINKED ISSUE #505: reset_registries + interpreter exit segfaults in final GC on Linux
## Summary

`_hgraph.reset_registries()` followed by ordinary interpreter exit segfaults during final garbage collection on Linux. macOS happens to tolerate the identical sequence, which is why it went unnoticed until the PR #504 wheel-test leg ran the first plain script that resets and then exits.

## Reproduction

Isolated in an OrbStack ubuntu VM against a wheel-configured shared install (`-DSKBUILD=2 -DHGRAPH_BUILD_SHARED=ON -DHGRAPH_BUILD_PYTHON_BINDINGS=ON -DHGRAPH_ENABLE_PYTHON_USER_NODES=ON`, pyarrow-backed Arrow, gcc-14), with controls:

```python
import _hgraph, hgraph
from hgraph import TS, pass_through
from hgraph.test import eval_node

assert eval_node(pass_through, [1], resolution_dict={"ts": TS[int]}) == [1]
_hgraph.reset_registries()          # <- with this line: SIGSEGV at exit
# without the reset: clean exit
```

Crash signature (faulthandler):

```
Fatal Python error: Segmentation fault
Current thread ... (most recent call first):
  Garbage-collecting
  <no Python frame>
```

Deleting every script-local object and running `gc.collect()` before the reset does not help — the crash is not the script's objects.

## Analysis

After `reset_registries()`, every nanobind-wrapped native handle created before the reset — including the `hgraph` package's own import-time objects (materialized schemas, subscription-cache entries keyed by earlier generations, module-level handles) — points into the freed interned-metadata generation. Final interpreter teardown garbage-collects those wrappers; their C++ destructors dereference the freed metadata. glibc has returned the pages, so Linux faults; macOS's allocator keeps them mapped, so the same UB passes silently.

This is the documented test-only reset hazard (`registry_reset.h`: "Reset is a test-only facility; production processes never reset") surfacing at teardown. The pytest suites that reset between every test somehow exit along a safe teardown path; a plain script exit does not.

## Current mitigation

`tests/python_extension_consumer/check.py` asserts everything, reports success, and leaves via `os._exit(0)`, skipping the hazardous teardown (landed on the checkpoint-3 branch, PR #504). Any future test script that resets and then exits must do the same.

## Proper fix (to schedule deliberately, not patch ad hoc)

Make stale-generation handle destruction safe, e.g.:

- generation-stamp the nb-wrapped handle payloads and no-op the destructor when `TypeRegistry::reset_generation()` has moved past their stamp; or
- have `reset_registries()` flip a process flag that turns subsequent native handle destructors into leaks (reset is test-only, so leaking the prior generation at exit is sound); or
- keep freed metadata generations alive until process exit (arena the interned storage per generation).

Whichever shape is chosen should come with a Linux regression test: the reproduction script above run as a subprocess asserting exit code 0 **without** `os._exit`.

Related memory of the investigation: PR #504 review thread; the OrbStack repro recipe (wheel-style flags + `~/venv312` + `build-linux-repro`).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### PR DESCRIPTION
**Pull request by hhenson**
Fixes #505. **Still reproduced on current `main`** (`bd9f7f1e1`) before this change — I checked rather than assumed, since the issue is ten days old.

## Diagnosis

The stack named it precisely:

```
#0  0x00007fff0000007f in ?? ()      <- call through a freed function pointer
#1  _hgraph.abi3.so                   <- our destructor
#2  libnanobind-abi3.so
#3  subtype_dealloc
#6  local_clear                       <- a threading.local dict
#7  local_dealloc
#10 gc_collect_main / finalize_modules
```

`GlobalState.instance()` caches a native-backed store on a `threading.local` (`_state.py:176-179`). It outlives the reset; the final GC at interpreter exit collects it, and its destructor dispatches through a plan the reset freed. Linux has returned the pages and faults; macOS keeps them mapped and the same UB passes silently.

Deleting that one cached object before the reset makes the crash disappear — that was the confirmation before writing any fix.

## The fix

This is the reset contract's own rule reaching the Python layer. `registry_reset.h` already requires that any cache holding pointers into another registry's artifacts be cleared *before* its lender frees them — that is why the native teardown is ordered. The Python caches were simply outside that sequence.

`reset_registries()` now runs registered hooks first, and `_state.py` registers one for its thread-local.

**A hook, not self-invalidation.** `registry_reset.h` blesses caches that compare `reset_generation()` and drop themselves on next access, and that is the wrong shape here: a script that resets and exits never makes that access, and the stale wrapper is exactly what exit-time GC then collects.

## Evidence

Rebuilt the wheel-configured shared build on `hg-linux` — the configuration that originally reproduced it — before and after:

| | before | after |
|---|---|---|
| repro (reset + ordinary exit) | **SIGSEGV, exit 139** | **exit 0** |
| control (no reset) | exit 0 | exit 0 |
| new regression tests | — | 3 passed |

The tests assert a **subprocess exit code**, as the issue requested, and carry no `os._exit`. That shape is not incidental: an in-process assertion cannot observe this, and macOS cannot observe it at all — which is why it went unnoticed originally.

## Two things I did not do

- **`tests/python_extension_consumer/check.py` still ends in `os._exit(0)`.** That mitigation is probably now unnecessary, but its comment cites hazards beyond the thread-local (decoration-time `@graph` handles), and verifying removal needs the installed-SDK consumer fixture built on Linux — a separate exercise. Worth a follow-up; I would rather leave a working mitigation than remove it on inference.
- **`_published_contexts` in `_core.py`** also holds natives (ports and type handles). It is a balanced push/pop stack, so it should be empty outside a wiring context, and I did not add a speculative hook for it. The mechanism is there if a case turns up.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: python/tests/test_registry_reset_teardown.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+
+def test_reset_twice_then_exit_does_not_crash():
+    """A second reset must not resurrect the hazard through a re-cached state."""
+    source = RESET_THEN_EXIT.replace(
+        '_hgraph.reset_registries()\n    print("ok")',
+        '_hgraph.reset_registries()\n'
+        '    assert eval_node(pass_through, [2], resolution_dict={"ts": TS[int]}) == [2]\n'
+        '    _hgraph.reset_registries()\n'
+        '    print("ok")',
```
[the remark this answers is not in the document]
Comment (hhenson) [reply python/tests/test_registry_reset_teardown.py:56]: Correct, and it was worse than an untested path — the test was silently a duplicate of the first one.

`RESET_THEN_EXIT` is dedented at definition, so the replacement keyed on `'...reset_registries()\n    print("ok")'` matched nothing, `source` came back unchanged, and the case ran a single reset. It passed, which is exactly why it went unnoticed.

Rewritten as its own dedented script rather than a patch over the first, and asserted: the twice-script now contains two `reset_registries()` calls. Thanks for catching it.

File: python/hgraph/_wiring/_state.py [hunk 2]
```diff
@@ -6,6 +6,26 @@
...
+    through, so the wrapper has to die while they are still alive -- otherwise
+    it is collected at interpreter exit and calls through freed memory, which
+    is a segfault on Linux and silent UB on macOS (issue #505).
+
+    Registered as a reset hook rather than left to self-invalidate on next
+    access: a script that resets and exits never makes that access, and the
+    stale wrapper is exactly what teardown then collects.
+    """
+    for attribute in list(vars(_global_state_local)):
+        delattr(_global_state_local, attribute)
```
[the remark this answers is not in the document]
Comment (hhenson) [reply python/hgraph/_wiring/_state.py:25]: Correct, and it is why this PR is being closed in favour of #569 rather than patched.

`vars(_global_state_local)` sees only the calling thread, so a worker that called `instance()` keeps its state through the reset and crashes when it later exits. Tracking cached states across threads would work, but it treats the symptom: the reason a worker can hold a stale state at all is that `instance()` lazily created and cached one for anyone who asked.

#569 removes that. `instance()` throws when no scope is active, and the thread-local is written only by `GlobalContext.__enter__`/`__exit__` -- I checked every write on that branch. A thread can therefore hold a state only *while inside a live scope*, never as a stale cache, so the cross-thread hazard has no source. Resetting while another thread sits in an active context is resetting under a live wiring, which is out of contract either way.

Verified on hg-linux: the #505 repro exits 0 on that branch with no hook present at all.

**Comment by hhenson** [comment]
Superseded by #569, which fixes #505 at the root.

This PR cleared the Python-side cache at reset time. #569 removes the reason the cache existed: `GlobalState.instance()` lazily created and cached a state for any caller — including hgraph's own code, 39 times across 10 files — so the compat accessor's side effect became the lifecycle. With `instance()` throwing and the thread-local written only by `GlobalContext.__enter__`/`__exit__`, wiring leaves nothing behind and a reset has nothing of the previous generation to strand.

That also answers both review findings here: the cross-thread hazard has no source once nothing can cache lazily, and the broken double-reset test is fixed and carried across.

Verified on hg-linux: the repro exits 0 against a wheel built from #569 with `register_reset_hook` confirmed absent. The subprocess regression tests have moved to #569.
